Session time zones

zvonand opened the pull request:
Continues #41197 , closes #34600

**TODO: REVIEW TCP PROTOCOL REVISION NR BEFORE MERGE**

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add setting `session_timezone`, it is used as default timezone for session when not explicitly specified.

alexey-milovidov commented:
Ok, but it can slow down the queries.
Let's do it like this:
- the setting will be named `default_timezone`;
- the corresponding DateLUT will be put into the query context and initialized somewhere near the creation of the query context;
- the `instance` method will obtain the reference of the already prepared DateLUT object if there is one; or create it as usual, otherwise.

PS. I like the idea, finally, we will be able to implement it!

zvonand commented:
@alexey-milovidov 
> the setting will be named `default_timezone`
I think we'd better leave `implicit_timezone`. It is not really "default", this name can be misleading. Maybe even `implicit_session_timezone`?

zvonand committed:
upd

zvonand committed:
re-add tests

zvonand commented:
Currently we have strange things in logs if the setting is enabled:
```
2023.01.30 16:22:02.936630 [ 8350 ] {} <Debug> DNSResolver: Updating DNS cache
.....
2023.01.30 20:22:04.539752 [ 8268 ] {072aba11-6413-4432-ab34-049b9f2a3ce2} <Debug> executeQuery: (from 127.0.0.1:39390) select 1 (stage: Complete)
.....
2023.01.30 16:24:24.278801 [ 8269 ] {657d3ca7-b555-4c75-b6f8-db6d77b693f3} <Debug> TCPHandler: Processed in 0.000749333 sec.
.....
2023.01.30 16:22:17.936768 [ 8399 ] {} <Debug> DNSResolver: Updating DNS cache
```
Not really strange -- I understand where they come from. Looking into this RN

zvonand committed:
do another way and logs work

zvonand committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into zvonand-implicit-tz

zvonand committed:
update test names

zvonand committed:
add docs + tiny cleanup

den-crane commented:
```sql
SELECT timeZone(), timeZoneOf(now()) SETTINGS implicit_timezone = 'Asia/Novosibirsk'
┌─timeZone()────┬─timeZoneOf(now())─┐
│ Europe/Berlin │ Asia/Novosibirsk  │
└───────────────┴───────────────────┘
```

It's actually bad idea. It will break applications and drivers. Because they are allow to set any `SETTINGS` (or settings can be set in users profiles. And they use `timeZone()` function to deduce the default timezone (of types w/o tz). 
`timeZone()` must return `Asia/Novosibirsk`

zvonand commented:
> `timeZone()` must return `Asia/Novosibirsk`

Not sure, see this:
[/src/Functions/ServerConstants.cpp L](https://github.com/ClickHouse/ClickHouse/blob/5154b04cfb8a87ec577b11bcd63b4b91fe9a709f/src/Functions/serverConstants.cpp#L63)

den-crane commented:
Also I don't understand what this PR does.

```sql
select now() format TSV;
2023-02-22 21:47:31

select now() format TSV;
2023-02-22 21:47:44 ???????????
```

Also I don't see any impact on inserts / text parsing.

den-crane commented:
```
    /// Returns the server time zone.
    class FunctionTimezone 
```

default Timezone of the data coming from the server.

When I query

```
select timezone() format TSV;
Etc/UTC

select now() x format TSVWithNamesAndTypes;
x
DateTime
2023-02-22 22:03:53
```
I now that timezone of the server is Etc/UTC, and `x` has a type DateTime with Etc/UTC timezone.

den-crane commented:
> > When I query ....
> 
> I don't get what you're trying to do there

I am explaining how JDBC/ODBC/Python/.net drivers use `select timezone();` call.

zvonand commented:
If I do `insert into test_tz  values (2, toDateTime('2000-01-01 00:00:00'));` it works fine. 
I'll look into it also, but AFAIR it does this best-effort conversion (your insert queries) to the destination datatype, including timezone. Are you sure this is incorrect behavior?

zvonand commented:
Ok, thanks for pointing it out. Will cleanup and come back soon

den-crane commented:
I see. Part of the problem is in the client

```
clickhouse-client --implicit_timezone=Asia/Novosibirsk -q 'select now()'
2023-02-22 22:30:51

select now() settings implicit_timezone = 'Asia/Novosibirsk' format TSV ;
2023-02-22 22:31:31

curl "http://localhost:8123?implicit_timezone=Asia/Novosibirsk&query=select+now()+format+TSV"
2023-02-23 05:32:43
```

den-crane commented:
One more issue

```sql
drop table test_tz;
create table test_tz ( d DateTime) Engine=Memory as select toDateTime('2000-01-01 00:00:00', 'UTC');

OK
------
select * from test_tz where d = toDateTime('2000-01-01 00:00:00');
0 rows in set. Elapsed: 0.001 sec.  

NOT OK
----
select * from test_tz where d = '2000-01-01 00:00:00';
┌───────────────────d─┐
│ 2000-01-01 00:00:00 │   -- 0 rows expected
└─────────────────────┘

curl "http://localhost:8123?implicit_timezone=Asia/Novosibirsk&query=select+*+from+test_tz+where+d='2000-01-01T00:00:00'"
2000-01-01 00:00:00        -- 0 rows expected
```

zvonand commented:
> I see. Part of the problem is in the client

Not actually the client. My mistake -- I omitted global context when creating DateLUT. WiP

den-crane commented:
@alexey-milovidov about the naming

psql `set timezone TO 'GMT';`
oracle `ALTER SESSION SET TIME_ZONE='Canada/Eastern';`
mysql `SET @@session.time_zone = "+00:00";`

alexey-milovidov commented:
@den-crane Yes, I like to name it just `timezone`; let's do it.

alexey-milovidov commented:
The code LGTM, the only concerns are around performance, but we shouldn't have places where DateLUT::instance is in the inner loop...

alexey-milovidov commented:
> And they use timeZone() function to deduce the default timezone (of types w/o tz).
timeZone() must return Asia/Novosibirsk

I agree. Interesting, why doesn't it work already as expected?
According to the code, everything in the query context should be affected by the setting.

alexey-milovidov commented:
> for i=2 text '2000-01-01 00:00:00' should be parsed as 'Asia/Novosibirsk' ( 946663200 )

+1, I expect it to work this way...

zvonand committed:
Added standalone function to get server's own timezone

Fix missing global_context

zvonand committed:
Added docs for new serverTimeZone function
Updated tests and docs

den-crane commented:
Fast test for some reason thinks that `timezone` function is not documented, though it's an alias to  `timeZone`.

zvonand commented:
Hmm, strange. 

BTW, I think it's a good idea to add some instructions on creating these inline docs. E.g. there: https://clickhouse.com/docs/en/development/developer-instruction

It is quite a non-obvious thing, IMO it deserves a line in contributors instruction :)

zvonand commented:
Also, DateTime is being built differently in common cases (even without having this new setting set up):
```sql
CREATE TABLE test_tz2 (`i` Int64, `d` DateTime('Asia/Novosibirsk')) ENGINE = Memory

INSERT INTO test_tz2 VALUES (1, toDateTime('2000-01-01 00:00:00'));

INSERT INTO test_tz2 VALUES (2, '2000-01-01 00:00:00');

SELECT * FROM test_tz2 format TSV

1	2000-01-01 05:00:00
2	2000-01-01 00:00:00
```

Is it ok and expected that they are treated differently?

den-crane commented:
>Is it ok and expected that they are treated differently? 

Yes, it's totally correct.  
`toDateTime('2000-01-01 00:00:00')` function parses it with server timezone.
Insert  parses `'2000-01-01 00:00:00'` it with 'Asia/Novosibirsk'.

zvonand commented:
Does it mean that this example was not good? https://github.com/ClickHouse/ClickHouse/pull/44149#issuecomment-1440888536

den-crane commented:
this example is  good.

```
-- d column is define as DateTime without timezone, it means that default timezone will be used during parsing of text

create table test_tz(i Int64, d DateTime) Engine=Memory;

-- some default timezone is used during insertion
-- may be for a test we should stabilize it by   `set implicit_timezone = 'UTC';`

insert into test_tz  values (1, '2000-01-01 00:00:00');

-- other default timezone is defined during insertion
set implicit_timezone = 'Asia/Novosibirsk';

-- this insert should parse a string using  'Asia/Novosibirsk'
insert into test_tz  values (2, '2000-01-01 00:00:00');
```

Here is an example with current CH:
```sql

zvonand committed:
update

zvonand commented:
IDK why fasttest is failing, everything is documented:
https://s3.amazonaws.com/clickhouse-test-reports/44149/c3a6efe0310ec23521eb97c1b53c6616f72ba7a0/fast_test.html

den-crane commented:
@zvonand you need to remove `timezone` line from ./tests/queries/0_stateless/02415_all_new_functions_must_be_documented.reference

This test checks which functions can be not documented. 
And it contains  `timezone`.

zvonand committed:
update undocumented funcs reference

den-crane commented:
@zvonand you need to fix merge conflicts

zvonand committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into zvonand-implicit-tz

den-crane commented:
timezone setting is not propagated to the clickhouse-client which renders and parses strings too.

```sql
set timezone='Canada/Atlantic';

select  toDateTime(1640995200) format TSV;
2022-01-01 00:00:00

select toString(toDateTime(1640995200)) format TSV;
2021-12-31 20:00:00
```

den-crane commented:
https://github.com/ClickHouse/ClickHouse/blob/49fb028498b5aedd22dd8d31cc82b3af5230ecc7/src/Server/TCPHandler.cpp#L1257

```
grep -R 'instance().getTimeZone' *
programs/server/Server.cpp:    LOG_TRACE(log, "Initialized DateLUT with time zone '{}'.", DateLUT::instance().getTimeZone());
programs/keeper/Keeper.cpp:    LOG_TRACE(log, "Initialized DateLUT with time zone '{}'.", DateLUT::instance().getTimeZone());
src/Interpreters/executeQuery.cpp:        .timezone = DateLUT::instance().getTimeZone(),
src/Server/GRPCServer.cpp:        *result.mutable_time_zone() = DateLUT::instance().getTimeZone();
src/Server/TCPHandler.cpp:        writeStringBinary(DateLUT::instance().getTimeZone(), *out);
src/Functions/serverConstants.cpp:        explicit FunctionTimezone(ContextPtr context) : FunctionConstantBase(String{DateLUT::instance().getTimeZone()}, context->isDistributed()) {}
src/Client/ClientBase.cpp:    const auto local_tz = DateLUT::instance().getTimeZone();
```

zvonand committed:
upd

den-crane commented:
@alexey-milovidov probably timezone is sent using Native protocol during initialization (hello) only. I guess need to modify Native protocol and send timezone with each query result.

zvonand commented:
IMO its not the only option. checking now, let's see if it works

zvonand committed:
update

zvonand commented:
@den-crane i don't think it will be hard to do. Shall I?

zvonand committed:
tcp protocol modification (min revision to be updated)

zvonand committed:
typo fix

den-crane commented on src/Core/ProtocolDefines.h:
```diff
@@ -72,3 +72,5 @@
 
 #define DBMS_MIN_PROTOCOL_VERSION_WITH_PASSWORD_COMPLEXITY_RULES 54461
+
+#define DBMS_MIN_PROTOCOL_VERSION_WITH_TIMEZONE_UPDATES 54461
```
copy / paste error ?
expected 54461 ?

zvonand replied:
Not erro, but placeholder. I dont know yet what will be the final revision - master changes constantly and rapidly :) I think its to be changed right before merge

zvonand committed:
fix receive of timezone update on processing

zvonand committed:
increment protocol version

zvonand committed:
revert protocol changes, found better way

zvonand committed:
upd

zvonand committed:
fix

zvonand committed:
fix style

zvonand committed:
update tests + exception

den-crane commented:
>We can modify client's settings just like server's.

How you can know that a server has changed a setting? I guess you cannot.

>There is no way to handle that with packets from server.

Why?

zvonand commented:
> There is no way to handle that with packets from server.

Let me re-check that. Had some issue yesterday.

zvonand committed:
Revert "fix style"

This reverts commit a7a3c9d1a675743e776fde32c96ccd9bbfc94e46.

zvonand committed:
Revert "revert protocol changes, found better way"

This reverts commit 3a918ae66a984451e0db0f56ffa6232b897ad62f.

zvonand committed:
added validation on setting modification

zvonand commented:
Work is still in progress here. Last problem I am working on is a delay in setting update via native protocol, see:
1. Launch everything and do `set timezone = 'New/Tz'`.
2. After executing that query, server sends client a packet with timezone, but it is a timezone that was there prior to modification. So, next query will be processed by client as if the timezone was unchanged.
3. After performing another successful query, server sends the correct (updated) timezone. 
If you modify the setting again, this delay will persist.

zvonand committed:
Merge branch 'master' into zvonand-implicit-tz

zvonand committed:
Merge branch 'master' into zvonand-implicit-tz

den-crane commented:
@zvonand fast test is failed. https://s3.amazonaws.com/clickhouse-test-reports/44149/a871c36d382919fb09a21bd268e91a1a671f3282/fast_test.html

```
2023-03-17 13:32:48 Received exception from server (version 23.3.1):
2023-03-17 13:32:48 Code: 36. DB::Exception: Received from localhost:9000. DB::Exception: Invalid time zone: Абырвалг. (BAD_ARGUMENTS)
2023-03-17 13:32:48 (query: SET timezone = 'Абырвалг';)

```

zvonand commented:
> fast test is failed

I know, it's just an outdated test. But this is not an actual issue, now looking at the stuff described above

Enmk commented on src/Core/SettingsFields.h:
```diff
@@ -545,6 +546,43 @@ void SettingFieldMultiEnum<EnumT, Traits>::readBinary(ReadBuffer & in)
+    cctz::time_zone validated_tz;
+    void validateTimezone(const std::string & str)
+    {
+        if (str != "" && !cctz::load_time_zone(str, &validated_tz))
```
Please check against `extern const char * auto_time_zones[];`, for details see https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/System/StorageSystemTimeZones.cpp#L7

zvonand replied:
Done, see latest commit

Enmk commented on src/DataTypes/TimezoneMixin.h:
```diff
@@ -15,7 +15,7 @@ class TimezoneMixin
     explicit TimezoneMixin(const String & time_zone_name = "")
         : has_explicit_time_zone(!time_zone_name.empty())
-        , time_zone(DateLUT::instance(time_zone_name))
+        , time_zone(time_zone_name.empty() ? DateLUT::instance() : DateLUT::instance(time_zone_name))
```
Why?

Enmk replied:
I would rather add `DateLUT::serverTimezone()` function and use it explicitly from corresponding places. Also, is `TimeZoneMixin` used inside logs? I thought that logging and other lower-level things (like TTL) it use `DateLUT` directly...

Enmk commented on src/Server/TCPHandler.cpp:
```diff
@@ -452,10 +452,14 @@ void TCPHandler::runImpl()
 
+                            if (client_tcp_protocol_version >= DBMS_MIN_PROTOCOL_VERSION_WITH_TIMEZONE_UPDATES
+                                && client_tcp_protocol_version >= DBMS_MIN_REVISION_WITH_SERVER_TIMEZONE)
+                                sendTimezone();
```
This is too late, maybe send timezone in `query_context->setInputInitializer` together with an empty header block?
https://github.com/ClickHouse/ClickHouse/blob/master/src/Server/TCPHandler.cpp#L335

zvonand replied:
Let's try and see 🙃

Enmk replied:
OR you can find out where header block is sent and send the timezone before or right after that.
Like https://github.com/ClickHouse/ClickHouse/blob/master/src/Server/TCPHandler.cpp#L810

Enmk commented on src/Common/DateLUT.h:
```diff
@@ -16,22 +17,50 @@
+                    return date_lut.getImplementation(effective_time_zone);
+            }
+
+            const auto global_context = DB::CurrentThread::get().getGlobalContext();
```
In which case global context may have overridden timezone ?

Enmk commented on src/Common/DateLUT.h:
```diff
@@ -16,22 +17,50 @@
+    /// If it is not set, server's timezone (the one which server has) is being used.
     static ALWAYS_INLINE const DateLUTImpl & instance()
     {
+        std::string effective_time_zone;
```
TODO: move inside next `if`-block

Enmk reviewed (changes requested):
Some changes required.
Also please specify how lower-level cases of timezone usage (e.g. merges, partitioninng, TTLs, logs, etc.) make sure that true server timezone is used.

Enmk committed:
Hackish way of setting up timezone on the client

Warning: lots of debug logging

zvonand committed:
Merge pull request #1 from Enmk/setting_timezone_on_client

Hackish way of setting up timezone on the client

zvonand committed:
fixed delay

zvonand committed:
moved getting server TZ DateLUT to separate place, upd tests and fix

zvonand committed:
resolve

zvonand committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into zvonand-implicit-tz

zvonand committed:
fix according to updates in ThreadStatus.h

zvonand committed:
remove additional logging

zvonand committed:
tryfix fasttest

zvonand committed:
upd remotequeryexecutor on receive timezone

zvonand committed:
fix style

zvonand committed:
update Timezone packet processing

zvonand committed:
update timezone packet handling in remote inserter

zvonand committed:
fix error on connection drop after 1st query

zvonand committed:
optimize

zvonand committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into zvonand-implicit-tz

zvonand committed:
renamed setting

zvonand committed:
fix linking issue

zvonand committed:
Lighter timezone validation

Reused external variable from src/Storages/System/StorageSystemTimeZones.generated.cpp
Required changes to CMakeLists of some standalone modules to link
properly

zvonand committed:
add errorcode reference

zvonand committed:
fix keeper standalone linking

zvonand committed:
tidy

Enmk commented on docs/en/sql-reference/functions/date-time-functions.md:
````diff
@@ -43,6 +43,25 @@ Alias: `timezone`.
+**Syntax**
+
+``` sql
+timeZone()
````
serverTimeZone() ?

Enmk replied:
```suggestion
serverTimeZone()
```

Enmk commented on docs/en/operations/settings/settings.md:
```diff
@@ -4008,6 +4008,35 @@ Default value: `0`.
+
+Possible values:
+
+-    Any valid timezone in `Region/Place` notation, e.g. `Europe/Berlin`
```
```suggestion
-    Any timezone name from `system.time_zones`, e.g. `Europe/Berlin`, `UTC` or `Zulu`
```

Enmk commented on docs/en/operations/settings/settings.md:
```diff
@@ -4008,6 +4008,35 @@ Default value: `0`.
 
+## session_timezone {#session_timezone}
+
+If specified, sets an implicit timezone (instead of server-default). All DateTime/DateTime64 values (and/or functions results) that have no explicit timezone specified are treated as having this timezone instead of default.
```
Setting this to `''` (empty string), effectively resets implicit timezone to server timezone.

Enmk commented on programs/client/Client.cpp:
```diff
@@ -320,6 +322,13 @@ try
 
+    {
+        // All that just to set DB::CurrentThread::get().getGlobalContext()
+        // which is required for client timezone (pushed as from server) to work.
```
```suggestion
        // which is required for client timezone (pushed from server) to work.
```

zvonand committed:
small updates due to review

zvonand committed:
Update docs/en/operations/settings/settings.md

Co-authored-by: Vasily Nemkov <V.Nemkov@gmail.com>

zvonand committed:
Update docs/ru/operations/settings/settings.md

Enmk commented on src/Core/SettingsFields.h:
```diff
@@ -2,6 +2,7 @@
 
 #include <Poco/Timespan.h>
 #include <Poco/URI.h>
+//#include <cctz/time_zone.h>
```
Cleanup:
```suggestion
```

zvonand committed:
Update src/Core/SettingsFields.h

Co-authored-by: Vasily Nemkov <V.Nemkov@gmail.com>

zvonand committed:
Update src/Client/ClientBase.cpp

Co-authored-by: Vasily Nemkov <V.Nemkov@gmail.com>

zvonand committed:
update cmakelists

zvonand committed:
cleanup cmakelists

zvonand committed:
style

zvonand committed:
Merge branch 'master' into zvonand-implicit-tz

zvonand commented:
Latest test fails seems irrelevant, they weren't there at previous commit

den-crane commented:
The issue with implicit conversion still exists.

```sql
clickhouse-local

select timezone();
┌─timezone()─┐
│ Etc/UTC    │
└────────────┘

create table test_tz ( d DateTime) Engine=Memory as select toDateTime('2000-01-01 00:00:00', 'UTC');

select * from test_tz where d = toDateTime('2000-01-01 00:00:00') settings session_timezone ='Asia/Novosibirsk';
0 rows in set. Elapsed: 0.016 sec. --- correct result

select * from test_tz where d = '2000-01-01 00:00:00' settings session_timezone ='Asia/Novosibirsk';
┌───────────────────d─┐
│ 2000-01-01 00:00:00 │ --- incorrect result, 0 rows expected
└─────────────────────┘ 
```

Let's compare with the current behaviour

```sql
 cat  /etc/clickhouse-server/config.d/t.xml
<clickhouse>
    <timezone>Asia/Novosibirsk</timezone>
</clickhouse>

select timezone();
┌─timezone()───────┐
│ Asia/Novosibirsk │
└──────────────────┘

create table test_tz ( d DateTime) Engine=Memory as select toDateTime('2000-01-01 00:00:00', 'UTC');

select * from test_tz where d = toDateTime('2000-01-01 00:00:00');
0 rows in set. Elapsed: 0.001 sec.   --- correct result

select * from test_tz where d = '2000-01-01 00:00:00';
0 rows in set. Elapsed: 0.001 sec.  --- correct result
```

zvonand commented:
@den-crane I cannot reproduce neither of the mentioned issues, e.g. see the first one:
```
SELECT timezone()
┌─timezone()────┐
│ Europe/Berlin │
└───────────────┘


CREATE TABLE test_tz
(
    `d` DateTime
)
ENGINE = Memory AS
SELECT toDateTime('2000-01-01 00:00:00', 'UTC')


SELECT *
FROM test_tz
WHERE d = toDateTime('2000-01-01 00:00:00')
SETTINGS session_timezone = 'Asia/Novosibirsk'

0 rows in set. Elapsed: 0.007 sec. 


SELECT *
FROM test_tz
WHERE d = '2000-01-01 00:00:00'
SETTINGS session_timezone = 'Asia/Novosibirsk'

0 rows in set. Elapsed: 0.007 sec.
```

Did you build it from scratch or used binaries from the PR CI/CD?

zvonand committed:
First portion of speedup

Do not generate DateLUT for each serialized and deserialized
date

robot-clickhouse-ci-1 commented:
<!-- automatic status comment for PR #44149 from zvonand/ClickHouse:zvonand-implicit-tz -->
This is an automated comment for commit a218f010e85c0bc995df17f5416bee4bcd8e6148 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/44149/a218f010e85c0bc995df17f5416bee4bcd8e6148/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🟡 pending</td></tr>
</table>

zvonand committed:
add timezone param

zvonand committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into zvonand-implicit-tz

zvonand committed:
speedup vol. II

zvonand committed:
fix stateless

zvonand committed:
fix Date32

zvonand committed:
cleanup

zvonand committed:
Merge branch 'master' into zvonand-implicit-tz

zvonand committed:
Merge branch 'master' into zvonand-implicit-tz

zvonand committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into zvonand-implicit-tz

Enmk commented on src/Client/ClientBase.cpp:
```diff
@@ -1075,6 +1080,13 @@ void ClientBase::onProgress(const Progress & value)
+{
+    Settings settings;
+    settings.session_timezone = tz;
+    global_context->applySettingsChanges(settings.changes());
```
why not just `Context::setSetting` ? It doesn't require the `Settings` instance

Enmk commented on src/Client/Connection.cpp:
```diff
@@ -355,6 +355,10 @@ void Connection::receiveHello()
             nonce.emplace(read_nonce);
         }
     }
+    else if (packet_type == Protocol::Server::TimezoneUpdate)
```
Maybe consider not sending it from server side on hello?

zvonand committed:
update settings from master

zvonand committed:
upd

rschu1ze commented on src/IO/ReadHelpers.h:
```diff
@@ -706,7 +706,7 @@ inline void convertToDayNum(DayNum & date, ExtendedDayNum & from)
 
 template <typename ReturnType = void>
-inline ReturnType readDateTextImpl(DayNum & date, ReadBuffer & buf)
+inline ReturnType readDateTextImpl(DayNum & date, ReadBuffer & buf, const DateLUTImpl & date_lut = DateLUT::instance())
```
Functions `readDateTextImpl()` are helper methods and not supposed to be called from outside file ReadHelpers.h. We should make argument `date_lut` mandatory (the little bit of extra convenience is not needed).

rschu1ze commented on src/IO/ReadHelpers.h:
```diff
@@ -1159,8 +1159,11 @@ inline bool tryReadText(IPv6 & x, ReadBuffer & buf) { return tryReadIPv6Text(x,
-inline void readText(LocalDate & x, ReadBuffer & buf) { readDateText(x, buf); }
+
 inline void readText(DayNum & x, ReadBuffer & buf) { readDateText(x, buf); }
+inline void readText(DayNum & x, ReadBuffer & buf, const DateLUTImpl & time_zone) { readDateText(x, buf, time_zone); }
```
Could we have just one function (maybe with default argument) instead of overloads l. 1163 and l. 1164.

rschu1ze reviewed (commented):
Which issue does this PR fix?

den-crane commented:
I downloaded 23.5.1.1144.

```
-rw-rw-r-- 1 ubuntu ubuntu     53588 May 10 00:16 clickhouse-client_23.5.1.1144_arm64.deb
-rw-rw-r-- 1 ubuntu ubuntu 188497994 May 10 00:16 clickhouse-common-static_23.5.1.1144_arm64.deb
-rw-rw-r-- 1 ubuntu ubuntu     78364 May 10 00:16 clickhouse-server_23.5.1.1144_arm64.deb
```

see my test https://github.com/ClickHouse/ClickHouse/pull/49795/commits/af8da326bc71effd20eadf8f520451fbcabfcc3c

```sql
create table test_tz (d DateTime) Engine=Memory as select toDateTime('2000-01-01 00:00:00', 'UTC');

select *, timezone() from test_tz where d = toDateTime('2000-01-01 00:00:00') settings session_timezone ='Asia/Novosibirsk';

0 rows in set. Elapsed: 0.015 sec.


select *, timezone() from test_tz where d = '2000-01-01 00:00:00' settings session_timezone ='Asia/Novosibirsk';
┌───────────────────d─┬─timezone()───────┐
│ 2000-01-01 00:00:00 │ Asia/Novosibirsk │
└─────────────────────┴──────────────────┘
```

zvonand commented on docs/en/operations/settings/settings.md:
```diff
@@ -4034,6 +4034,36 @@ Default value: `0`.
 
+## session_timezone {#session_timezone}
+
+If specified, sets an implicit timezone (instead of server-default). All DateTime/DateTime64 values (and/or functions results) that have no explicit timezone specified are treated as having this timezone instead of default.
```
Why? I think it is a good one. It is really implicit e.g. in `toDateTime('2022-12-12 12:12:12.11', 2)`, it is not set explicitly

zvonand commented on src/IO/ReadHelpers.h:
```diff
@@ -1172,6 +1175,10 @@ template <typename T>
 requires is_arithmetic_v<T>
 inline void readQuoted(T & x, ReadBuffer & buf) { readText(x, buf); }
 
+template <typename T>
```
Not a good idea. We need this timezone only for a narrow scope of types. E.g. we don't need it for Int, but having a default parameter would mean constructing an unnecessary and expensive object

zvonand commented on src/IO/ReadHelpers.h:
```diff
@@ -1248,6 +1259,24 @@ inline void readCSVSimple(T & x, ReadBuffer & buf)
 }
 
+// standalone overload for dates: to avoid instantiating DateLUTs while parsing other types
+template <typename T>
```
see the message above,
https://github.com/ClickHouse/ClickHouse/pull/44149#discussion_r1191690360

zvonand commented on src/Core/SettingsFields.cpp:
```diff
@@ -451,6 +461,23 @@ String SettingFieldEnumHelpers::readBinary(ReadBuffer & in)
+
+void SettingFieldTimezone::validateTimezone(std::string_view str)
+{
+    if (!str.empty() && !checkIsExistingTimeZone(str))
```
Why? Not good to allow wrong setting in case we can easily avoid it. This check is cheap and is not performed regularly

zvonand commented on tests/queries/0_stateless/02681_timezone_setting.sql:
```diff
@@ -0,0 +1,11 @@
+SET session_timezone = 'Абырвалг'; -- { serverError BAD_ARGUMENTS}
+
+SET session_timezone = 'Asia/Novosibirsk';
+SELECT toDateTime64(toDateTime64('2022-12-12 23:23:23.123', 3), 3, 'Europe/Zurich');
```
This is exactly what is being tested here

zvonand committed:
update docs

zvonand committed:
fixed docs 2

den-crane commented:
CI also fails https://s3.amazonaws.com/clickhouse-test-reports/49795/99c7fe74443955047ef7bca68df37ddbd7f11798/fast_test.html

zvonand commented:
@den-crane well, I am now not sure this behavior is _not really_ wrong
When we compare like this: `== toDateTime(...)`, DateTime is being created from scratch using the new `session_timezone` setting.
But when we do `d == '2000-......`, this string is being parsed to same DateTime kind as table column. So, in your case it is casted to the same timezone as the DateTime in table, and no wonder the query works this way. 
Implicit String-to-Something works like this for all data types. So, generally this behavior is _correct_.

zvonand commented:
I didn't find this difference in docs, but it definitely should be here. But not sure where to put this

den-crane commented:
@zvonand see https://github.com/ClickHouse/ClickHouse/pull/44149#issuecomment-1520795014

zvonand commented:
@den-crane yes, my concern was exactly about the first part of the comment (the one with implicit conversion in WHERE part)

den-crane commented:
current behaviour / current expectations

## column without TZ
```
select timezone();
┌─timezone()─┐
│ Etc/UTC    │
└────────────┘

create table test ( A DateTime ) Engine = Memory as select 1577836800;

select count() from test where A = toDateTime('2020-01-01 00:00:00');
┌─count()─┐
│       1 │
└─────────┘

select count() from test where A ='2020-01-01 00:00:00';
┌─count()─┐
│       1 │
└─────────┘

select count() from test where A = toDate('2020-01-01');
┌─count()─┐
│       1 │
└─────────┘

select count() from test where A = '2020-01-01';
┌─count()─┐
│       1 │
└─────────┘
```

## column with TZ

```
create table test ( A DateTime('Europe/Berlin' ) ) Engine = Memory as select 1577836800;

select count() from test where A = toDateTime('2020-01-01 00:00:00');
┌─count()─┐
│       1 │
└─────────┘

select count() from test where A ='2020-01-01 00:00:00';
┌─count()─┐
│       0 │
└─────────┘

select count() from test where A = toDate('2020-01-01');
┌─count()─┐
│       0 │
└─────────┘

select count() from test where A = '2020-01-01';
┌─count()─┐
│       0 │
└─────────┘
```

I expect the same behaviour if the server timezone is different from UTC and overridden to UTC by the new session_timezone setting.
If behaviour is different it should be documented.

zvonand committed:
fix conflict and update

zvonand committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into zvonand-implicit-tz

zvonand committed:
resolve ambiguity by adding a section to docs

zvonand committed:
fix incode docs

zvonand commented on src/Common/DateLUT.h:
```diff
@@ -16,22 +17,56 @@
+            /// Timezone is passed in query_context, but on CH-Client we have no query context,
+            /// and each time we modify client's global context
+            const auto global_context = DB::CurrentThread::get().getGlobalContext();
+            if (global_context)
```
Why? `session_timezone` being an empty string means it is not set. Thus we need to check global_context (see comment)

zvonand commented on programs/copier/ClusterCopierApp.cpp:
```diff
@@ -44,7 +44,7 @@ void ClusterCopierApp::initialize(Poco::Util::Application & self)
     auto curr_pid = Poco::Process::id();
 
-    process_id = std::to_string(DateLUT::instance().toNumYYYYMMDDhhmmss(timestamp)) + "_" + std::to_string(curr_pid);
+    process_id = std::to_string(DateLUT::serverTimezoneInstance().toNumYYYYMMDDhhmmss(timestamp)) + "_" + std::to_string(curr_pid);
```
yes, seems so, it took a while. I would be grateful for another view

zvonand commented on src/Core/SettingsFields.h:
```diff
@@ -545,6 +545,47 @@ void SettingFieldMultiEnum<EnumT, Traits>::readBinary(ReadBuffer & in)
 
+/* Setting field for specifying user-defined timezone. It is basically a string, but it needs validation.
+ */
+struct SettingFieldTimezone
```
Settings.h, l. 750 -- where the `session_timezone` session is defined

zvonand commented on programs/keeper/CMakeLists.txt:
```diff
@@ -107,6 +107,8 @@ if (BUILD_STANDALONE_KEEPER)
         ${CMAKE_CURRENT_SOURCE_DIR}/../../src/Daemon/GraphiteWriter.cpp
         ${CMAKE_CURRENT_BINARY_DIR}/../../src/Daemon/GitHash.generated.cpp
 
+        ${CMAKE_CURRENT_BINARY_DIR}/../../src/Storages/System/StorageSystemTimeZones.generated.cpp
```
TBH I don't remember where exactly, but it failed to compile and link without it.

zvonand commented on src/CMakeLists.txt:
```diff
@@ -192,6 +192,11 @@ list (APPEND dbms_headers
+# Required for validation of Timezone in session_timezone setting.
+# This way we don't need to create timezone via cctz each time, but check against pregenerated char**
+list (APPEND dbms_sources
+        Storages/System/StorageSystemTimeZones.generated.cpp)
```
see contrib/cctz-cmake/CMakeLists.txt

zvonand committed:
update

zvonand committed:
fix client thread attachment + make better tests

zvonand committed:
Merge branch 'master' into zvonand-implicit-tz

zvonand committed:
Merge branch 'master' into zvonand-implicit-tz

zvonand committed:
remove unused import

zvonand committed:
Merge branch 'master' into zvonand-implicit-tz

zvonand committed:
cosmetic changes

zvonand committed:
Merge branch 'zvonand-implicit-tz' of github.com:zvonand/ClickHouse into zvonand-implicit-tz

zvonand committed:
update docs spelling check failed

zvonand committed:
Merge branch 'master' into zvonand-implicit-tz

zvonand committed:
upd spell

zvonand commented:
@rschu1ze I did some updates, including those according to your past review

rschu1ze commented:
Great, thanks. I'll try to take a look today or tomorrow.

zvonand commented:
My only concern here now is this commit: https://github.com/ClickHouse/ClickHouse/pull/44149/commits/1e8c0a2db9671f0862975499f16b923a49c3a2ec
I did it to make tz setting validation easier, but it required linking an additional file to almost all targets.
It works fine, and previous one worked fine (though it was heavier). Which is better here? esp. if we don't really expect user to change the setting frequently

rschu1ze commented on src/Core/SettingsFields.h:
```diff
@@ -545,6 +545,47 @@ void SettingFieldMultiEnum<EnumT, Traits>::readBinary(ReadBuffer & in)
+    String value;
+    bool changed = false;
+
+//    explicit SettingFieldTimezone(std::string_view str = {}) { validateTimezone(std::string(str)); value = str; }
```
(Too much commented out code in this file)

rschu1ze reviewed (commented):
Lgtm, thanks. Only a few minor comments, then we can merge.

zvonand commented on programs/client/Client.cpp:
```diff
@@ -320,6 +322,14 @@ try
     initTtyBuffer(toProgressOption(config().getString("progress", "default")));
 
+    {
+        // All that just to set DB::CurrentThread::get().getGlobalContext()
```
yes, kind of :) it was the only good option, thanks @Enmk for help

zvonand commented on docs/en/operations/settings/settings.md:
```diff
@@ -4203,6 +4203,62 @@ Default value: `0`.
 Use this setting only for backward compatibility if your use cases depend on old syntax.
 :::
 
+## session_timezone {#session_timezone}
```
there is already a setting called `timezone` in `config.xml`. This may also introduce some mess. 
Also, don't remember exactly why, but I **_had_** to change the name, I used to call it `timezone`, but there were some errors. Maybe we can come up with a better naming option :) 
maybe something like `user_defined_timezone` or `user_timezone` would be fine

rschu1ze replied:
Yes, setting `timezone` in "config.xml" *is* the global/server-level timezone.

As far as I see, there is no precedence for an equally named global/session setting, but I don't think anything can mess up. It would be very appreciated if you could make a try to rename `session_timezone` to `timezone` in a separate PR, after this one is merged.

(`user_defined_timezone` / `user_timezone` ar a bit too chatty for my taste)

zvonand replied:
yes, agree

zvonand replied:
It actually would be nicer to rename server's setting to `server_timezone`, but I guess it is a really bad option

rschu1ze replied:
It is a bad idea :-) (and unfortunately settings can't be aliased as of now, so we can't deprecate setting "timezone").

zvonand replied:
Ok, after this is merged I can try to change it to `timezone`

zvonand commented on docs/en/operations/settings/settings.md:
```diff
@@ -4203,6 +4203,62 @@ Default value: `0`.
+
+If specified, sets an implicit timezone (instead of [server default](../server-configuration-parameters/settings.md#server_configuration_parameters-timezone).
+All DateTime/DateTime64 values (and/or functions results) that have no explicit timezone specified are treated as having this timezone instead of default.
+A value of `''` (empty string) configures the session timezone to the server default timezone.
```
Well, checking if string is empty is cheap, easy and obvious, and it also doesn't break the way things currently work, while empty string is passed by default, e.g. in TimezoneMixin and all its children.
This would introduce additional checks and string comparisons here and there, not good for performance also.
Moreover, in all other settings emptiness means that setting is not set and default behavior is working.
IMO when one wants to change back to default, empty string looks more default than some string with empty value, especially if he has used other settings previously

rschu1ze replied:
Makes sense, thanks.

zvonand commented on docs/en/operations/settings/settings.md:
```diff
@@ -4203,6 +4203,62 @@ Default value: `0`.
+CREATE TABLE test_tz (`d` DateTime('UTC')) ENGINE = Memory AS SELECT toDateTime('2000-01-01 00:00:00', 'UTC');
+
+SELECT *, timezone() FROM test_tz WHERE d = toDateTime('2000-01-01 00:00:00') SETTINGS session_timezone = 'Asia/Novosibirsk'
+0 rows in set.
```
Yes, exactly like this

zvonand commented on tests/queries/0_stateless/02737_timezone_setting.sql:
```diff
@@ -0,0 +1,27 @@
+
+-- test simple queries
+SELECT toDateTime(toDateTime('2022-12-12 23:23:23'), 'Europe/Zurich');
+SELECT toDateTime64(toDateTime64('2022-12-12 23:23:23.123', 3), 3, 'Europe/Zurich') SETTINGS session_timezone = 'America/Denver';
```
This is to show that overriding is working properly (at some point it actually wasn't while working on this PR 😃 )
Also this test can be a reference for some use cases

zvonand committed:
changes requested in review

zvonand committed:
merge master and resolve conflict

zvonand committed:
fix style

zvonand committed:
Update 02737_timezone_setting.sql

zvonand committed:
Update settings.md

zvonand commented on docs/en/operations/settings/settings.md:
````diff
@@ -4203,6 +4203,62 @@ Default value: `0`.
+Examples:
+
+```sql
+SELECT timeZone(), serverTimezone() FORMAT TSV
````
added

zvonand committed:
Update settings.md

zvonand commented:
@rschu1ze Updated according to your review, and some answers/explanations left in comments

rschu1ze committed:
Update settings.md

rschu1ze committed:
Update settings.md

rschu1ze committed:
Update settings.md

rschu1ze committed:
Update settings.md

rschu1ze commented:
> My only doubt here now is this commit: [1e8c0a2](https://github.com/ClickHouse/ClickHouse/commit/1e8c0a2db9671f0862975499f16b923a49c3a2ec)
> I did it to make tz setting validation easier, but it required linking an additional file to almost all targets.
> It works fine, and previous one worked fine (though it was heavier). Which is better here?
> (also, seems like `session_timezone` isn't a thing that would be modified frequently)

So `SettingFieldTimezone::validateTimezone()` is only called for statements `SET session_timezone = ...` or `SELECT ... FROM ... SETTINGS session_timezone = ...`, right? (btw., I think the former statement will be used more frequently in practice). Tbh, optimizing the check seems a bit overkill to me- even if the setting is applied within a `SELECT` it is a one-time constant cost and amortizes with the query duration. So, so to keep things simple and not build `StorageSystemTimeZones.generated.cpp` into all binaries, I'd prefer to revert this optimization.

zvonand commented:
>  I'd prefer to revert this optimization.
Sure, will do now

zvonand committed:
revert tz validation optimization

zvonand committed:
Update SettingsFields.cpp

zvonand committed:
typo in SettingsFields.h

zvonand commented:
Timezone validation is reverted, as soon as the tests finish, it's ready from my side

zvonand commented:
Timezone validation is reverted, current test fails look unrelated
@rschu1ze

zvonand commented:
@rschu1ze are there any other concerns about this PR?

alexey-milovidov commented:
Tests have failed.

zvonand commented:
ok, I see some test fixes merged recently (i.e.  `02703_max_local_write_bandwidth`, `02561_null_as_default_more_formats`, `02783_parsedatetimebesteffort_syslog`).

zvonand committed:
Merge branch 'master' into zvonand-implicit-tz

zvonand committed:
small

rschu1ze commented:
@zvonand Thanks - I'll check/merge tomorrow (am ooo today).

zvonand commented:
I honestly see no connection between fails in my last few commits and my changes -- the fails before and after merging master are also completely different 
pretty much the same amount of random fails (stateless, integration) are in recent master commits, e.g.  [5867f39](https://github.com/ClickHouse/ClickHouse/commit/5867f39fc7fc5649a52d670ddb863d12ff313865), and some tests are just flaky, i.e. [01443_merge_truncate_long](https://github.com/ClickHouse/ClickHouse/issues/50294)

alexey-milovidov commented:
@rschu1ze Please fix the `01443_merge_truncate_long` test before merging.

zvonand committed:
Merge branch 'master' into zvonand-implicit-tz

zvonand commented:
Stateless tests (release, analyzer):
* `02772_s3_crash` - removed by #51239
* `00992_system_parts_race_condition_zookeeper_long` -- flaky in master, timeout hit, #50434

zvonand commented:
Test run finished again, see test fails analysis in the comment above.
`01443_merge_truncate_long` did not fail this time, and it is unrelated (and affected many other (already merged) PRs)

(Had to update to newer master due to merge conflicts)

@rschu1ze

zvonand committed:
Merge branch 'master' into zvonand-implicit-tz

zvonand committed:
Update Settings.h

zvonand commented:
Only fail now is `00159_parallel_formatting_json_and_friends`, which I could not find the cause of.
Looks mostly like a random timeout. Maybe there could be some connection in case of a real fail, but it is just a timeout.

zvonand commented:
I once again reviewed the current fail and previous ones and they are not related to changes proposed in this PR

zvonand commented:
@rschu1ze could you please close the issue #34600 as completed by this PR?
(It wasn't linked to PR but it is directly related)

azat commented on src/Core/Settings.h:
```diff
@@ -770,6 +770,7 @@ class IColumn;
     M(Bool, allow_experimental_undrop_table_query, false, "Allow to use undrop query to restore dropped table in a limited time", 0) \
     M(Bool, keeper_map_strict_mode, false, "Enforce additional checks during operations on KeeperMap. E.g. throw an exception on an insert for already existing key", 0) \
     M(UInt64, extract_kvp_max_pairs_per_row, 1000, "Max number pairs that can be produced by extractKeyValuePairs function. Used to safeguard against consuming too much memory.", 0) \
+    M(Timezone, session_timezone, "", "The default timezone for current session or query. The server default timezone if empty.", 0) \
```
Hm, there is `use_client_time_zone` (with underscore), maybe it worth to add an alias for `use_client_timezone` as well to comply the convention?

Although it seems that `time_zone` and `timezone` has a more or less the same distribution:

```
 gg time_zone  | wc -l
1069
$ gg timezone  | wc -l
987

$ gg '\<timezone\>'  | wc -l
643
$ gg '\<time_zone\>'  | wc -l
848
```

den-crane commented:
For some reason I was sure that X-ClickHouse-Timezone reflects session_timezone.
But it does not 
* https://github.com/ClickHouse/ClickHouse/issues/63359
